File: scripts/engines/go/steamdb_engine.go
package main

import (
	"encoding/json"
	"fmt"
	htmlstd "html"
	"io"
	"net/http"
	"os"
	"regexp"
	"strconv"
	"strings"
	"time"
)

var (
	reTags        = regexp.MustCompile(`(?is)<[^>]+>`)
	reWhitespace  = regexp.MustCompile(`\s+`)
	reTitle       = regexp.MustCompile(`(?is)<title[^>]*>(.*?)</title>`)
	reH1          = regexp.MustCompile(`(?is)<h1[^>]*>(.*?)</h1>`)
	reTableRow    = regexp.MustCompile(`(?is)<tr[^>]*>\s*<t[dh][^>]*>(.*?)</t[dh]>\s*<t[dh][^>]*>(.*?)</t[dh]>\s*</tr>`)
	reDigit       = regexp.MustCompile(`\d+`)
	reRating      = regexp.MustCompile(`(?is)([\d.]+)%.*?(\d[\d,]*)\s+reviews`)
	reDashSplit   = regexp.MustCompile(`\s*[–-]\s*`)
	reSteamSuffix = regexp.MustCompile(`(?i)\s*-\s*SteamDB.*$`)
)

type sizePattern struct {
	re     *regexp.Regexp
	isFull bool
}

var sizePatterns = []sizePattern{
	{re: regexp.MustCompile(`(?i)Total\s+size\s+on\s+disk\s+is\s+([\d.]+)\s*(GiB|MiB|GB|MB)`), isFull: true},
	{re: regexp.MustCompile(`(?i)total\s+download\s+size\s+is\s+([\d.]+)\s*(GiB|MiB|GB|MB)`), isFull: false},
	{re: regexp.MustCompile(`(?is)([\d.]+)\s*(GiB|MiB|GB|MB).*?total`), isFull: false},
	{re: regexp.MustCompile(`(?is)<td>Size</td>\s*<td[^>]*>([\d.]+)\s*(GiB|MiB|GB|MB)`), isFull: false},
	{re: regexp.MustCompile(`(?i)Disk\s+Space[:\s]+([\d.]+)\s*(GiB|MiB|GB|MB)`), isFull: false},
}

func stripTags(value string) string {
	out := reTags.ReplaceAllString(value, " ")
	out = htmlstd.UnescapeString(out)
	out = reWhitespace.ReplaceAllString(out, " ")
	return strings.TrimSpace(out)
}

func isBotProtectionPage(page string, title string) bool {
	snippet := strings.ToLower(title + "\n" + page)
	if len(snippet) > 10000 {
		snippet = snippet[:10000]
	}

	markers := []string{
		"checking your browser",
		"just a moment",
		"attention required",
		"cloudflare",
		"cf-chl",
		"security check to access",
	}

	for _, marker := range markers {
		if strings.Contains(snippet, marker) {
			return true
		}
	}
	return false
}

func parseSteamDBHTML(page string) map[string]any {
	info := map[string]any{}

	var title string
	if h1Match := reH1.FindStringSubmatch(page); len(h1Match) > 1 {
		title = stripTags(h1Match[1])
	}
	if title == "" {
		if titleMatch := reTitle.FindStringSubmatch(page); len(titleMatch) > 1 {
			title = stripTags(titleMatch[1])
			title = strings.TrimSpace(reSteamSuffix.ReplaceAllString(title, ""))
		}
	}
	if title != "" {
		if isBotProtectionPage(page, title) {
			return nil
		}
		info["name"] = title
	}

	for _, row := range reTableRow.FindAllStringSubmatch(page, -1) {
		if len(row) < 3 {
			continue
		}
		label := stripTags(row[1])
		value := stripTags(row[2])
		if label == "" || value == "" {
			continue
		}

		labelLower := strings.ToLower(label)
		switch {
		case strings.Contains(labelLower, "developer"):
			info["developer"] = value
		case strings.Contains(labelLower, "publisher"):
			info["publisher"] = value
		case strings.Contains(labelLower, "release date"):
			info["releaseDate"] = value
		case strings.Contains(labelLower, "last record update"):
			parts := reDashSplit.Split(value, 2)
			if len(parts) > 0 {
				info["lastUpdate"] = strings.TrimSpace(parts[0])
			}
		case strings.Contains(labelLower, "dlc"):
			if digit := reDigit.FindString(value); digit != "" {
				if parsed, err := strconv.Atoi(digit); err == nil {
					info["dlcCount"] = parsed
				}
			}
		}
	}

	for _, pattern := range sizePatterns {
		match := pattern.re.FindStringSubmatch(page)
		if len(match) < 3 {
			continue
		}

		size, err := strconv.ParseFloat(match[1], 64)
		if err != nil || size <= 0 || size >= 2000 {
			continue
		}

		unit := strings.ToUpper(strings.TrimSpace(match[2]))
		isGB := strings.Contains(unit, "GIB") || unit == "GB"
		if isGB {
			info["size"] = size * 1024 * 1024 * 1024
		} else {
			info["size"] = size * 1024 * 1024
		}

		info["sizeFormatted"] = fmt.Sprintf("%v %s", size, strings.ReplaceAll(unit, "I", ""))
		if pattern.isFull {
			info["sizeType"] = "FULL"
		} else {
			info["sizeType"] = "Base"
		}
		break
	}

	if rating := reRating.FindStringSubmatch(page); len(rating) > 2 {
		info["rating"] = fmt.Sprintf("%s%%", rating[1])
		info["reviewCount"] = strings.ReplaceAll(rating[2], ",", "")
	}

	if len(info) == 0 {
		return nil
	}
	return info
}

func main() {
	if len(os.Args) < 2 {
		fmt.Println("null")
		os.Exit(1)
	}

	appID := strings.TrimSpace(os.Args[1])
	if appID == "" {
		fmt.Println("null")
		os.Exit(1)
	}

	timeoutMs := 15000
	if len(os.Args) >= 3 {
		if parsed, err := strconv.Atoi(os.Args[2]); err == nil && parsed > 0 {
			timeoutMs = parsed
		}
	}

	url := fmt.Sprintf("https://steamdb.info/app/%s/", appID)
	client := &http.Client{Timeout: time.Duration(timeoutMs) * time.Millisecond}

	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		fmt.Println("null")
		os.Exit(2)
	}
	req.Header.Set("User-Agent", "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36")

	resp, err := client.Do(req)
	if err != nil {
		fmt.Println("null")
		os.Exit(3)
	}
	defer resp.Body.Close()

	body, err := io.ReadAll(resp.Body)
	if err != nil {
		fmt.Println("null")
		os.Exit(4)
	}

	info := parseSteamDBHTML(string(body))
	if info == nil {
		fmt.Println("null")
		return
	}

	encoded, err := json.Marshal(info)
	if err != nil {
		fmt.Println("null")
		os.Exit(5)
	}

	fmt.Println(string(encoded))
}

File: scripts/engines/steamdb_engine.py
#!/usr/bin/env python3
import html
import json
import re
import sys
import urllib.error
import urllib.request


def strip_tags(value: str) -> str:
  cleaned = re.sub(r"<[^>]+>", " ", value or "", flags=re.IGNORECASE | re.DOTALL)
  cleaned = html.unescape(cleaned)
  cleaned = re.sub(r"\s+", " ", cleaned).strip()
  return cleaned


def parse_float(value: str):
  try:
    return float(value)
  except Exception:
    return None


def parse_int(value: str):
  if not value:
    return None
  match = re.search(r"\d+", value)
  if not match:
    return None
  try:
    return int(match.group(0))
  except Exception:
    return None


def is_bot_protection_page(page: str, title: str = "") -> bool:
  snippet = f"{title}\n{(page or '')[:10000]}".lower()
  markers = [
    "checking your browser",
    "just a moment",
    "attention required",
    "cloudflare",
    "cf-chl",
    "security check to access",
  ]
  return any(marker in snippet for marker in markers)


def parse_steamdb_html(page: str):
  info = {}

  h1_match = re.search(r"<h1[^>]*>(.*?)</h1>", page, flags=re.IGNORECASE | re.DOTALL)
  title_match = re.search(r"<title[^>]*>(.*?)</title>", page, flags=re.IGNORECASE | re.DOTALL)

  title = ""
  if h1_match:
    title = strip_tags(h1_match.group(1))
  if not title and title_match:
    title = strip_tags(title_match.group(1))
    title = re.sub(r"\s*-\s*SteamDB.*$", "", title, flags=re.IGNORECASE).strip()
  if is_bot_protection_page(page, title):
    return None
  if title:
    info["name"] = title

  row_pattern = re.compile(
    r"<tr[^>]*>\s*<t[dh][^>]*>(.*?)</t[dh]>\s*<t[dh][^>]*>(.*?)</t[dh]>\s*</tr>",
    flags=re.IGNORECASE | re.DOTALL,
  )

  for row_match in row_pattern.finditer(page):
    label = strip_tags(row_match.group(1))
    value = strip_tags(row_match.group(2))
    if not label or not value:
      continue

    label_lower = label.lower()
    if "developer" in label_lower:
      info["developer"] = value or "Unknown"
    elif "publisher" in label_lower:
      info["publisher"] = value or "Unknown"
    elif "release date" in label_lower:
      info["releaseDate"] = value
    elif "last record update" in label_lower:
      info["lastUpdate"] = re.split(r"\s*[–-]\s*", value, maxsplit=1)[0].strip()
    elif "dlc" in label_lower:
      parsed_dlc = parse_int(value)
      if parsed_dlc is not None:
        info["dlcCount"] = parsed_dlc

  size_patterns = [
    (re.compile(r"Total\s+size\s+on\s+disk\s+is\s+([\d.]+)\s*(GiB|MiB|GB|MB)", re.IGNORECASE), True),
    (re.compile(r"total\s+download\s+size\s+is\s+([\d.]+)\s*(GiB|MiB|GB|MB)", re.IGNORECASE), False),
    (re.compile(r"([\d.]+)\s*(GiB|MiB|GB|MB).*?total", re.IGNORECASE | re.DOTALL), False),
    (re.compile(r"<td>Size</td>\s*<td[^>]*>([\d.]+)\s*(GiB|MiB|GB|MB)", re.IGNORECASE | re.DOTALL), False),
    (re.compile(r"Disk\s+Space[:\s]+([\d.]+)\s*(GiB|MiB|GB|MB)", re.IGNORECASE), False),
  ]

  for pattern, is_full in size_patterns:
    match = pattern.search(page)
    if not match:
      continue

    size = parse_float(match.group(1))
    unit = (match.group(2) or "").upper()
    if size is None or size <= 0 or size >= 2000:
      continue

    is_gb = ("GIB" in unit) or unit == "GB"
    info["size"] = size * 1024 * 1024 * 1024 if is_gb else size * 1024 * 1024
    info["sizeFormatted"] = f"{size} {unit.replace('I', '')}"
    info["sizeType"] = "FULL" if is_full else "Base"
    break

  rating_match = re.search(r"([\d.]+)%.*?(\d[\d,]*)\s+reviews", page, flags=re.IGNORECASE | re.DOTALL)
  if rating_match:
    info["rating"] = f"{rating_match.group(1)}%"
    info["reviewCount"] = rating_match.group(2).replace(",", "")

  return info if len(info) > 0 else None


def fetch_steamdb_info(app_id: str, timeout_ms: int):
  url = f"https://steamdb.info/app/{app_id}/"
  req = urllib.request.Request(
    url,
    headers={
      "User-Agent": "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"
    },
  )

  timeout_seconds = max(1.0, float(timeout_ms) / 1000.0)
  with urllib.request.urlopen(req, timeout=timeout_seconds) as response:
    page = response.read().decode("utf-8", errors="replace")

  return parse_steamdb_html(page)


def main():
  if len(sys.argv) < 2:
    print("null")
    return 1

  app_id = str(sys.argv[1]).strip()
  if not app_id:
    print("null")
    return 1

  timeout_ms = 15000
  if len(sys.argv) >= 3:
    try:
      parsed_timeout = int(sys.argv[2])
      if parsed_timeout > 0:
        timeout_ms = parsed_timeout
    except ValueError:
      pass

  try:
    data = fetch_steamdb_info(app_id, timeout_ms)
    if data is None:
      print("null")
      return 0

    print(json.dumps(data, ensure_ascii=True))
    return 0
  except urllib.error.URLError:
    print("null")
    return 0
  except Exception as exc:
    print(f"{{\"error\":\"internal\",\"message\":{json.dumps(str(exc))}}}")
    return 3


if __name__ == "__main__":
  sys.exit(main())
